zebra-consensus/src/transaction.rs

```diff
@@ -403,4 +402,57 @@ where
             )),
         }
     }
+
+    fn verify_transparent_inputs_and_outputs(
+        request: &Request,
+        network: Network,
+        inputs: &[transparent::Input],
+        script_verifier: script::Verifier<ZS>,
+    ) -> Result<
+        FuturesUnordered<Pin<Box<dyn Future<Output = Result<(), BoxError>> + Send>>>,
+        TransactionError,
+    > {
+        let transaction = request.transaction();
+
+        if transaction.is_coinbase() {
+            check::coinbase_tx_no_prevout_joinsplit_spend(&transaction)?;
+
+            Ok(FuturesUnordered::new())
+        } else {
+            // feed all of the inputs to the script and shielded verifiers
+            // the script_verifier also checks transparent sighashes, using its own implementation
+            let cached_ffi_transaction = Arc::new(CachedFfiTransaction::new(transaction));
+            let known_utxos = request.known_utxos();
+            let upgrade = request.upgrade(network);
+
+            let script_checks = (0..inputs.len())
+                .into_iter()
+                .map(move |input_index| {
+                    let request = script::Request {
+                        upgrade,
+                        known_utxos: known_utxos.clone(),
+                        cached_ffi_transaction: cached_ffi_transaction.clone(),
+                        input_index,
+                    };
+
+                    script_verifier.clone().oneshot(request).boxed()
+                })
+                .collect();
+
+            Ok(script_checks)
+        }
+    }
+
+    async fn wait_for_checks(
+        mut checks: FuturesUnordered<Pin<Box<dyn Future<Output = Result<(), BoxError>> + Send>>>,
+    ) -> Result<(), TransactionError> {
+        // Wait for all asynchronous checks to complete
+        // successfully, or fail verification if they error.
+        while let Some(check) = checks.next().await {
+            tracing::trace!(?check, remaining = checks.len());
+            check?;
+        }
+
+        Ok(())
+    }
```

Comment:
We're currently calling `wait_for_checks` at the end of every transaction version.
Could we just run the code at the end of `call` instead?

That way, every transaction version would return a `Result<FuturesUnordered<_>, TransactionError>`, and we can cut out the boilerplate `extend/wait_for_checks` lines.

Reply:
Good idea! Done.